### ISSUE TITLE
NCCL peer access is not supported error

### ISSUE THREAD

**Issue by amogkam** [ray-project/ray_lightning#24]
just the training was about to start (nvidia-smi shows the GPU memory filling up), there's a new error
```
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1603729062494/work/torch/lib/c10d/ProcessGroupNCCL.cpp:784, unhandled cuda error, NCCL version 2.7.8
```

Starting the Python training script using `NCCL_IB_DISABLE=1 python tune.py` does not help.

Enabled debug messages using `NCCL_DEBUG="INFO" NCCL_IB_DISABLE=1 python tune.py`, and the following new messages appeared:

```
(pid=909597) RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1603729062494/work/torch/lib/c10d/ProcessGroupNCCL.cpp:784, unhandled cuda error, NCCL version 2.7.8
(pid=909608) z-pc:909608:910337 [0] NCCL INFO Channel 00/02 :    0   1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO Channel 01/02 :    0   1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=909608) z-pc:909608:910337 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] 1/-1/-1->0->-1|-1->0->1/-1/-1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO Channel 00 : 0[2d000] -> 1[2e000] via P2P/IPC
(pid=909608) 
(pid=909608) z-pc:909608:910337 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=909608) z-pc:909608:910337 [0] NCCL INFO transport.cc:68 -> 1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO init.cc:766 -> 1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO init.cc:840 -> 1
(pid=909608) z-pc:909608:910337 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
(pid=909611) z-pc:909611:910338 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=909611) z-pc:909611:910338 [0] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] -1/-1/-1->1->0|0->1->-1/-1/-1
(pid=909611) z-pc:909611:910338 [0] NCCL INFO Channel 00 : 1[2e000] -> 0[2d000] via P2P/IPC
(pid=909611) 
(pid=909611) z-pc:909611:910338 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=909611) z-pc:909611:910338 [0] NCCL INFO transport.cc:68 -> 1
(pid=909611) z-pc:909611:910338 [0] NCCL INFO init.cc:766 -> 1
(pid=909611) z-pc:909611:910338 [0] NCCL INFO init.cc:840 -> 1
(pid=909611) z-pc:909611:910338 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
2021-03-30 18:20:32,548 ERROR trial_runner.py:616 -- Trial train_model_190f1_00000: Error processing event.
$ nvidia-smi topo -m
	GPU0	GPU1	CPU Affinity	NUMA Affinity
GPU0	 X 	PHB	0-31		N/A
GPU1	PHB	 X 	0-31		N/A
```

_Originally posted by @nyxynyx_

**Comment by amogkam** [comment]
Upgraded Pytorch to 1.8.1, a similar NCCL error appears

ray::ImplicitFunc.train_buffered() (pid=928774, ip=192.168.1.159)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 248, in run
    self._entrypoint()
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 315, in entrypoint
    return self._trainable_func(self.config, self._status_reporter,
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 576, in _trainable_func
    output = fn()
  File "tune.py", line 54, in train_model
    trainer.fit(model, dm)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 499, in fit
    self.dispatch()
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 546, in dispatch
    self.accelerator.start_training(self)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 169, in start_training
    results = process_results(futures, queue)
  File "/home/x/ray_lightning/ray_lightning/util.py", line 103, in process_results
    ray.get(ready)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
    return func(*args, **kwargs)
ray.exceptions.RayTaskError(RuntimeError): ray::RayExecutor.execute() (pid=928769, ip=192.168.1.159)
  File "python/ray/_raylet.pyx", line 480, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 432, in ray._raylet.execute_task.function_executor
  File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 31, in execute
    return fn(*args, **kwargs)
  File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 223, in train_remote
    super(RayPlugin, self).new_process(
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp_spawn.py", line 157, in new_process
    self.configure_ddp()
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp_spawn.py", line 195, in configure_ddp
    self._model = DistributedDataParallel(
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 446, in __init__
    self._sync_params_and_buffers(authoritative_rank=0)
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 457, in _sync_params_and_buffers
    self._distributed_broadcast_coalesced(
  File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 1155, in _distributed_broadcast_coalesced
    dist._broadcast_coalesced(
RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1616554788289/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled cuda error, NCCL version 2.7.8
ncclUnhandledCudaError: Call to CUDA function failed.
Also, when using NCLL_DEBUG="INFO"

(pid=929792) RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1616554788289/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled cuda error, NCCL version 2.7.8
(pid=929792) ncclUnhandledCudaError: Call to CUDA function failed.
(pid=929783) z-pc:929783:930303 [0] NCCL INFO Channel 00/02 :    0   1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO Channel 01/02 :    0   1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=929783) z-pc:929783:930303 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] 1/-1/-1->0->-1|-1->0->1/-1/-1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO Channel 00 : 0[2d000] -> 1[2e000] via P2P/IPC
(pid=929783) 
(pid=929783) z-pc:929783:930303 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=929783) z-pc:929783:930303 [0] NCCL INFO transport.cc:68 -> 1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO init.cc:766 -> 1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO init.cc:840 -> 1
(pid=929783) z-pc:929783:930303 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
(pid=929779) z-pc:929779:930304 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=929779) z-pc:929779:930304 [0] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] -1/-1/-1->1->0|0->1->-1/-1/-1
(pid=929779) z-pc:929779:930304 [0] NCCL INFO Channel 00 : 1[2e000] -> 0[2d000] via P2P/IPC
(pid=929779) 
(pid=929779) z-pc:929779:930304 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=929779) z-pc:929779:930304 [0] NCCL INFO transport.cc:68 -> 1
(pid=929779) z-pc:929779:930304 [0] NCCL INFO init.cc:766 -> 1
(pid=929779) z-pc:929779:930304 [0] NCCL INFO init.cc:840 -> 1
(pid=929779) z-pc:929779:930304 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
Packages versions being used are now

pytorch 1.7.0
pytorch-lightning 1.2.5
ray 1.2.0
cudatoolkit 10.2.89

**Comment by amogkam** [comment]
@nyxynyx this library uses TCP initialization to create the process group (https://pytorch.org/docs/stable/distributed.html#tcp-initialization) which requires P2P communication between all your devices. Do you have P2P enabled between devices on your machine?

**Comment by nyxynyx** [comment]
@amogkam Thanks for organizing this question into its own issue.

Before using `ray_lightning`, I have been able to train Pytorch Lightning models using both GPU at the same time for the same model with DDP.

```py
trainer = pl.Trainer(
    gpus=-1,
    precision=16,
    accelerator="ddp",
)
```

I have since tried the following combinations, all encountered the same NCCL error when trying to use both GPUs using `ray_lightning`:
- Pytorch 1.7.0, CUDA 10.2.89
- Pytorch 1.8.1, CUDA 10.2.89
- Pytorch 1.8.1, CUDA 11.1.1

How do we check whether P2P is enabled between the two RTX 2070 super?

**Comment by amogkam** [comment]
@nyxynyx is this same error happening even when you are not using Tune? Standard Pytorch Lightning does not use TCP initialization so that could be why it has been working for you. You can confirm by running this small example from here https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#basic-use-case, except using NCCL and TCP initialization for dist.init_process_group

**Comment by nyxynyx** [comment]
@amogkam I have not tried using Ray on multiple GPUs and without Tune.

I've tried running the small example linked, and it works only with `gloo`, not with `nccl`. With `nccl`, it simply hangs and the script does not finish running.

Is it possible to use another method to run `ray_lightning` with DDP other than NCCL with TCP initialization?

Below is the code that I've adapted to use for my test

```py
import os
import sys
import tempfile
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.optim as optim
import torch.multiprocessing as mp

from torch.nn.parallel import DistributedDataParallel as DDP


def setup(rank, world_size):
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12355"

    # gloo: works
    # dist.init_process_group("gloo", rank=rank, world_size=world_size)

    # # nccl: hangs forever
    dist.init_process_group(
        "nccl", init_method="tcp://10.1.1.20:23456", rank=rank, world_size=world_size
    )


def cleanup():
    dist.destroy_process_group()


class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.net1 = nn.Linear(10, 10)
        self.relu = nn.ReLU()
        self.net2 = nn.Linear(10, 5)

    def forward(self, x):
        return self.net2(self.relu(self.net1(x)))


def demo_basic(rank, world_size):
    print(f"Running basic DDP example on rank {rank}.")
    setup(rank, world_size)

    # create model and move it to GPU with id rank
    model = ToyModel().to(rank)
    ddp_model = DDP(model, device_ids=[rank])

    loss_fn = nn.MSELoss()
    optimizer = optim.SGD(ddp_model.parameters(), lr=0.001)

    optimizer.zero_grad()
    outputs = ddp_model(torch.randn(20, 10))
    labels = torch.randn(20, 5).to(rank)
    loss_fn(outputs, labels).backward()
    optimizer.step()

    cleanup()


def run_demo(demo_fn, world_size):
    mp.spawn(demo_fn, args=(world_size,), nprocs=world_size, join=True)


if __name__ == "__main__":
    run_demo(demo_basic, 2)
```

**Comment by amogkam** [comment]
Ah interesting, so looks like this is more of a Pytorch/NCCL issue if the simple torch example is not working for you. I would recommend posting on their side if you want to use NCCL.

If you want to use gloo with GPU, you'd have to modify this line in the code https://github.com/ray-project/ray_lightning/blob/main/ray_lightning/ray_ddp.py#L238. 

I will push a patch to make the backend configurable. Hopefully this hack can unblock you for now.

**Comment by nyxynyx** [comment]
@amogkam I've managed to get the script working.

As soon as `setup()` was modified to remove `init_method="tcp://10.1.1.20:23456"` and instead perform TCP initialization using values from `os.environ["MASTER_ADDR"]`  and `os.environ["MASTER_PORT"] `

```py
def setup(rank, world_size):
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12355"

    dist.init_process_group("nccl", rank=rank, world_size=world_size)
```

the script started working!

Does this mean NCCL & TCP initialization are both working for my GPU devices?

**Comment by amogkam** [comment]
@nyxynyx Great, I'm glad it's working for you! So the change you made actually removed TCP initialization. If you don't provide an `init_method` to `dist.init_process_group` it uses Environment Variable initialization. You can read more about it here https://pytorch.org/docs/stable/distributed.html#tcp-initialization. So NCCL is working, but TCP initialization is not, I think because you might not have P2P communication enabled across your devices.

**Comment by amogkam** [comment]
But thanks for bringing this up! I think I'll make the change the default for this library to Environment Variable initialization since that seems to work on more setups and is the default used by Pytorch Lightning as well.

**Comment by nyxynyx** [comment]
@amogkam Thanks for the clarification! Looking forward to the changes :wink:

**Comment by amogkam** [comment]
Ah @nyxynyx now that I think about it, using env var initialization won't work when using with Ray Tune since you can have multiple distributed training runs running on the same node, and you don't each one to interfere with the environmnent variables with the other runs. Using TCP initialization is probably the safest bet.

In your case, env var initialization is not an issue since you are only running 1 Tune trial at a time (so there's only 1 process group ever created).

**Comment by nyxynyx** [comment]
@amogkam Thanks for working on the PR so quickly!

I tried using the latest fix that just merged in by running `git pull` then `pip install -e .`, but I still getting the same error. 

Is there an additional setting that we need to have in order to use env initialization with NCCL?

```py
$ git rev-parse HEAD
c13bbb19e949650c00917d243b1d78084743f700
```

**Error Stacktrace:**

```py
(pid=3126456) GPU available: True, used: True
(pid=3126456) TPU available: False, using: 0 TPU cores
(pid=3126456) Using native 16bit precision.
(pid=3126456) LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
(pid=3126454) Global seed set to 1234
(pid=3126454) initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/2
(pid=3126451) Global seed set to 1234
(pid=3126451) initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/2
(pid=3126451) z-pc:3126451:3126451 [0] NCCL INFO Bootstrap : Using [0]enp39s0:192.168.1.159<0> [1]br-850839b9f073:172.18.0.1<0> [2]veth1a84001:fe80::1886:45ff:febb:9007%veth1a84001<0>
(pid=3126451) z-pc:3126451:3126451 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
(pid=3126451) 
(pid=3126451) z-pc:3126451:3126451 [0] misc/ibvwrap.cc:63 NCCL WARN Failed to open libibverbs.so[.1]
(pid=3126451) z-pc:3126451:3126451 [0] NCCL INFO NET/Socket : Using [0]enp39s0:192.168.1.159<0> [1]br-850839b9f073:172.18.0.1<0> [2]veth1a84001:fe80::1886:45ff:febb:9007%veth1a84001<0>
(pid=3126451) z-pc:3126451:3126451 [0] NCCL INFO Using network Socket
(pid=3126454) z-pc:3126454:3126454 [0] NCCL INFO Bootstrap : Using [0]enp39s0:192.168.1.159<0> [1]br-850839b9f073:172.18.0.1<0> [2]veth1a84001:fe80::1886:45ff:febb:9007%veth1a84001<0>
(pid=3126454) z-pc:3126454:3126454 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
(pid=3126454) 
(pid=3126454) z-pc:3126454:3126454 [0] misc/ibvwrap.cc:63 NCCL WARN Failed to open libibverbs.so[.1]
(pid=3126454) z-pc:3126454:3126454 [0] NCCL INFO NET/Socket : Using [0]enp39s0:192.168.1.159<0> [1]br-850839b9f073:172.18.0.1<0> [2]veth1a84001:fe80::1886:45ff:febb:9007%veth1a84001<0>
(pid=3126454) z-pc:3126454:3126454 [0] NCCL INFO Using network Socket
(pid=3126454) NCCL version 2.7.8+cuda11.1
(pid=3126451) z-pc:3126451:3127583 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=3126451) z-pc:3126451:3127583 [0] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] -1/-1/-1->1->0|0->1->-1/-1/-1
(pid=3126454) z-pc:3126454:3127582 [0] NCCL INFO Channel 00/02 :    0   1
(pid=3126454) z-pc:3126454:3127582 [0] NCCL INFO Channel 01/02 :    0   1
(pid=3126454) z-pc:3126454:3127582 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=3126454) z-pc:3126454:3127582 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] 1/-1/-1->0->-1|-1->0->1/-1/-1
(pid=3126451) z-pc:3126451:3127583 [0] NCCL INFO Channel 00 : 1[2e000] -> 0[2d000] via P2P/IPC
(pid=3126454) z-pc:3126454:3127582 [0] NCCL INFO Channel 00 : 0[2d000] -> 1[2e000] via P2P/IPC
(pid=3126451) 
(pid=3126451) z-pc:3126451:3127583 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=3126451) z-pc:3126451:3127583 [0] NCCL INFO transport.cc:68 -> 1
(pid=3126451) z-pc:3126451:3127583 [0] NCCL INFO init.cc:766 -> 1
(pid=3126451) z-pc:3126451:3127583 [0] NCCL INFO init.cc:840 -> 1
(pid=3126451) z-pc:3126451:3127583 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
(pid=3126454) 
(pid=3126454) z-pc:3126454:3127582 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=3126454) z-pc:3126454:3127582 [0] NCCL INFO transport.cc:68 -> 1
(pid=3126454) z-pc:3126454:3127582 [0] NCCL INFO init.cc:766 -> 1
(pid=3126454) z-pc:3126454:3127582 [0] NCCL INFO init.cc:840 -> 1
(pid=3126454) z-pc:3126454:3127582 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
(pid=3126456) 2021-04-04 12:56:00,573   ERROR function_runner.py:254 -- Runner Thread raised error.
(pid=3126456) Traceback (most recent call last):
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 248, in run
(pid=3126456)     self._entrypoint()
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 315, in entrypoint
(pid=3126456)     return self._trainable_func(self.config, self._status_reporter,
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 576, in _trainable_func
(pid=3126456)     output = fn()
(pid=3126456)   File "train.py", line 311, in train
(pid=3126456)     trainer.fit(model, dm)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 499, in fit
(pid=3126456)     self.dispatch()
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 546, in dispatch
(pid=3126456)     self.accelerator.start_training(self)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
(pid=3126456)     self.training_type_plugin.start_training(trainer)
(pid=3126456)   File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 180, in start_training
(pid=3126456)     results = process_results(futures, queue)
(pid=3126456)   File "/home/x/ray_lightning/ray_lightning/util.py", line 103, in process_results
(pid=3126456)     ray.get(ready)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
(pid=3126456)     return func(*args, **kwargs)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/worker.py", line 1456, in get
(pid=3126456)     raise value.as_instanceof_cause()
(pid=3126456) ray.exceptions.RayTaskError(RuntimeError): ray::RayExecutor.execute() (pid=3126454, ip=192.168.1.159)
(pid=3126456)   File "python/ray/_raylet.pyx", line 480, in ray._raylet.execute_task
(pid=3126456)   File "python/ray/_raylet.pyx", line 432, in ray._raylet.execute_task.function_executor
(pid=3126456)   File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 39, in execute
(pid=3126456)     return fn(*args, **kwargs)
(pid=3126456)   File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 234, in train_remote
(pid=3126456)     super(RayPlugin, self).new_process(
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp_spawn.py", line 157, in new_process
(pid=3126456)     self.configure_ddp()
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp_spawn.py", line 195, in configure_ddp
(pid=3126456)     self._model = DistributedDataParallel(
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 446, in __init__
(pid=3126456)     self._sync_params_and_buffers(authoritative_rank=0)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 457, in _sync_params_and_buffers
(pid=3126456)     self._distributed_broadcast_coalesced(
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 1155, in _distributed_broadcast_coalesced
(pid=3126456)     dist._broadcast_coalesced(
(pid=3126456) RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1616554793803/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled cuda error, NCCL version 2.7.8
(pid=3126456) ncclUnhandledCudaError: Call to CUDA function failed.
(pid=3126456) Exception in thread Thread-2:
(pid=3126456) Traceback (most recent call last):
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/threading.py", line 932, in _bootstrap_inner
(pid=3126456)     self.run()
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 267, in run
(pid=3126456)     raise e
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 248, in run
(pid=3126456)     self._entrypoint()
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 315, in entrypoint
(pid=3126456)     return self._trainable_func(self.config, self._status_reporter,
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/tune/function_runner.py", line 576, in _trainable_func
(pid=3126456)     output = fn()
(pid=3126456)   File "train.py", line 311, in train
(pid=3126456)     trainer.fit(model, dm)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 499, in fit
(pid=3126456)     self.dispatch()
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 546, in dispatch
(pid=3126456)     self.accelerator.start_training(self)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
(pid=3126456)     self.training_type_plugin.start_training(trainer)
(pid=3126456)   File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 180, in start_training
(pid=3126456)     results = process_results(futures, queue)
(pid=3126456)   File "/home/x/ray_lightning/ray_lightning/util.py", line 103, in process_results
(pid=3126456)     ray.get(ready)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 47, in wrapper
(pid=3126456)     return func(*args, **kwargs)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/ray/worker.py", line 1456, in get
(pid=3126456)     raise value.as_instanceof_cause()
(pid=3126456) ray.exceptions.RayTaskError(RuntimeError): ray::RayExecutor.execute() (pid=3126454, ip=192.168.1.159)
(pid=3126456)   File "python/ray/_raylet.pyx", line 480, in ray._raylet.execute_task
(pid=3126456)   File "python/ray/_raylet.pyx", line 432, in ray._raylet.execute_task.function_executor
(pid=3126456)   File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 39, in execute
(pid=3126456)     return fn(*args, **kwargs)
(pid=3126456)   File "/home/x/ray_lightning/ray_lightning/ray_ddp.py", line 234, in train_remote
(pid=3126456)     super(RayPlugin, self).new_process(
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp_spawn.py", line 157, in new_process
(pid=3126456)     self.configure_ddp()
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/ddp_spawn.py", line 195, in configure_ddp
(pid=3126456)     self._model = DistributedDataParallel(
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 446, in __init__
(pid=3126456)     self._sync_params_and_buffers(authoritative_rank=0)
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 457, in _sync_params_and_buffers
(pid=3126456)     self._distributed_broadcast_coalesced(
(pid=3126456)   File "/opt/anaconda3/envs/test/lib/python3.8/site-packages/torch/nn/parallel/distributed.py", line 1155, in _distributed_broadcast_coalesced
(pid=3126456)     dist._broadcast_coalesced(
(pid=3126456) RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1616554793803/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled cuda error, NCCL version 2.7.8
(pid=3126456) ncclUnhandledCudaError: Call to CUDA function failed.
```

**Comment by amogkam** [comment]
Hmm it should be using env var initialization by default. Can you print out what the MASTER_ADDR and MASTER_PORT env cars are over here https://github.com/ray-project/ray_lightning/blob/main/ray_lightning/ray_ddp.py#L249. Also if you manually set these to what you had before 

os.environ["MASTER_ADDR"] = "localhost"
os.environ["MASTER_PORT"] = "12355"

then is it working for you?

**Comment by amogkam** [comment]
Also if you want to set any NCCL environment variables (like NCCL_DEBUG for example) you have to make sure that it is set in the worker processes as well. You can do so by passing in an init_hook to the plugin https://github.com/ray-project/ray_lightning/blob/main/ray_lightning/ray_ddp.py#L66 that sets these env vars.

**Comment by nyxynyx** [comment]
@amogkam At L249,

`os.getenv('MASTER_ADDR')` gives `192.168.1.159`
`os.getenv('MASTER_PORT')` gives `45439`

Using `ifconfig`, the IP address assigned to my ethernet interface is `192.168.1.159`.

If we set the env variables at L249, the following code in those lines

```py
        torch_backend = os.getenv("PL_TORCH_DISTRIBUTED_BACKEND")
        print('MASTER_ADDR:', os.getenv('MASTER_ADDR'))
        print('MASTER_PORT:', os.getenv('MASTER_PORT'))
        os.environ["MASTER_ADDR"] = "localhost"
        os.environ["MASTER_PORT"] = "12355"
        print('MASTER_ADDR 2:', os.getenv('MASTER_ADDR'))
        print('MASTER_PORT 2:', os.getenv('MASTER_PORT'))
```

gives the output

```py
(pid=3788394) initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/2
(pid=3788394) MASTER_ADDR: 192.168.1.159
(pid=3788394) MASTER_PORT: 53071
(pid=3788394) MASTER_ADDR 2: localhost
(pid=3788394) MASTER_PORT 2: 12355
(pid=3788410) MASTER_ADDR: 192.168.1.159
(pid=3788410) MASTER_PORT: 53071
(pid=3788410) MASTER_ADDR 2: localhost
(pid=3788410) MASTER_PORT 2: 12355
(pid=3788410) Global seed set to 1234
```

and the training starts with no NCCL errors.

**Comment by nyxynyx** [comment]
@amogkam 

Tried using the `init_hook` as suggested, 

```py
def train(config):

    # ...

    def init_hook():
        os.environ["NCCL_DEBUG"] = "INFO"

    trainer = pl.Trainer(
        gpus=-1,
        accelerator="ddp",
        plugins=[
            RayPlugin(num_workers=2, use_gpu=True, init_hook=init_hook),
        ],
        precision=16,
    )
```

here's the `NCCL INFO` messages:

```py
(pid=3817941) RuntimeError: NCCL error in: /opt/conda/conda-bld/pytorch_1616554793803/work/torch/lib/c10d/ProcessGroupNCCL.cpp:825, unhandled cuda error, NCCL version 2.7.8
(pid=3817941) ncclUnhandledCudaError: Call to CUDA function failed.
(pid=3817959) z-pc:3817959:3817959 [0] NCCL INFO Bootstrap : Using [0]enp39s0:192.168.1.159<0> [1]br-850839b9f073:172.18.0.1<0> [2]veth1a84001:fe80::1886:45ff:febb:9007%veth1a84001<0>
(pid=3817959) z-pc:3817959:3817959 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
(pid=3817959) 
(pid=3817959) z-pc:3817959:3817959 [0] misc/ibvwrap.cc:63 NCCL WARN Failed to open libibverbs.so[.1]
(pid=3817959) z-pc:3817959:3817959 [0] NCCL INFO NET/Socket : Using [0]enp39s0:192.168.1.159<0> [1]br-850839b9f073:172.18.0.1<0> [2]veth1a84001:fe80::1886:45ff:febb:9007%veth1a84001<0>
(pid=3817959) z-pc:3817959:3817959 [0] NCCL INFO Using network Socket
(pid=3817959) NCCL version 2.7.8+cuda11.1
(pid=3817959) z-pc:3817959:3819190 [0] NCCL INFO Channel 00/02 :    0   1
(pid=3817959) z-pc:3817959:3819190 [0] NCCL INFO Channel 01/02 :    0   1
(pid=3817959) z-pc:3817959:3819190 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=3817959) z-pc:3817959:3819190 [0] NCCL INFO Trees [0] 1/-1/-1->0->-1|-1->0->1/-1/-1 [1] 1/-1/-1->0->-1|-1->0->1/-1/-1
(pid=3817959) z-pc:3817959:3819190 [0] NCCL INFO Channel 00 : 0[2d000] -> 1[2e000] via P2P/IPC
(pid=3817924) z-pc:3817924:3817924 [0] NCCL INFO Bootstrap : Using [0]enp39s0:192.168.1.159<0> [1]br-850839b9f073:172.18.0.1<0> [2]veth1a84001:fe80::1886:45ff:febb:9007%veth1a84001<0>
(pid=3817924) z-pc:3817924:3817924 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
(pid=3817924) 
(pid=3817924) z-pc:3817924:3817924 [0] misc/ibvwrap.cc:63 NCCL WARN Failed to open libibverbs.so[.1]
(pid=3817924) z-pc:3817924:3817924 [0] NCCL INFO NET/Socket : Using [0]enp39s0:192.168.1.159<0> [1]br-850839b9f073:172.18.0.1<0> [2]veth1a84001:fe80::1886:45ff:febb:9007%veth1a84001<0>
(pid=3817924) z-pc:3817924:3817924 [0] NCCL INFO Using network Socket
(pid=3817924) z-pc:3817924:3819191 [0] NCCL INFO threadThresholds 8/8/64 | 16/8/64 | 8/8/64
(pid=3817924) z-pc:3817924:3819191 [0] NCCL INFO Trees [0] -1/-1/-1->1->0|0->1->-1/-1/-1 [1] -1/-1/-1->1->0|0->1->-1/-1/-1
(pid=3817924) z-pc:3817924:3819191 [0] NCCL INFO Channel 00 : 1[2e000] -> 0[2d000] via P2P/IPC
(pid=3817924) 
(pid=3817924) z-pc:3817924:3819191 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=3817924) z-pc:3817924:3819191 [0] NCCL INFO transport.cc:68 -> 1
(pid=3817924) z-pc:3817924:3819191 [0] NCCL INFO init.cc:766 -> 1
(pid=3817924) z-pc:3817924:3819191 [0] NCCL INFO init.cc:840 -> 1
(pid=3817924) z-pc:3817924:3819191 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
(pid=3817959) 
(pid=3817959) z-pc:3817959:3819190 [0] transport/p2p.cc:238 NCCL WARN failed to open CUDA IPC handle : 217 peer access is not supported between these two devices
(pid=3817959) z-pc:3817959:3819190 [0] NCCL INFO transport.cc:68 -> 1
(pid=3817959) z-pc:3817959:3819190 [0] NCCL INFO init.cc:766 -> 1
(pid=3817959) z-pc:3817959:3819190 [0] NCCL INFO init.cc:840 -> 1
(pid=3817959) z-pc:3817959:3819190 [0] NCCL INFO group.cc:73 -> 1 [Async thread]
2021-04-04 15:47:38,798 ERROR trial_runner.py:616 -- Trial train_7b610_00000: Error processing event.
```

**Comment by amogkam** [comment]
Thanks for the info @nyxynyx- I'm actually quite stumped on this one since the address being set looks to be correct. Can you try manually setting address to 192.168.1.159 and port to 12355, as well as address to localhost and port to 45439? I'm curious as to whether the address or the port is the problem (or both), but my hunch is that it's the address.

Also, just to confirm with localhost/12355 the `ToyModel` example from earlier is working fine right? If that example is not working with 192.168.1.159/45439 then it's most likely a Pytorch or NCCL error. I would recommend posting an issue on their repos as well with the ToyModel script. I'd be curious to find out what's going on as well.

In the meantime, hopefully either using gloo or manually setting localhost/12355 is working for you.

**Comment by nyxynyx** [comment]
@amogkam Thanks for continuing to dig into this issue! 

I can confirm that the ToyModel script is still working when we modify the setup function to 

```py
def setup(rank, world_size):
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12355"

    dist.init_process_group("nccl", rank=rank, world_size=world_size)
```

and when modified to 


```py
def setup(rank, world_size):
    os.environ["MASTER_ADDR"] = "192.168.1.159"
    os.environ["MASTER_PORT"] = "12355"

    dist.init_process_group("nccl", rank=rank, world_size=world_size)
```

Unfortunately, I am getting the same NCCL error when we modify the `init_ddp_connection` method in `test_ddp.py`  by inserting the following 

```py
os.environ["MASTER_ADDR"] = "192.168.1.159"
os.environ["MASTER_PORT"] = "45439"
```

or to 

```py
os.environ["MASTER_ADDR"] = "192.168.1.159"
os.environ["MASTER_PORT"] = "12355"
```

or to 

```py
os.environ["MASTER_ADDR"] = "localhost"
os.environ["MASTER_PORT"] = "45439"
```

or to

```py
os.environ["MASTER_ADDR"] = "localhost"
os.environ["MASTER_PORT"] = "12355"
```

This is mysterious indeed, I will be using gloo for now. Thanks!

**Comment by 2a3b4c** [comment]
@amogkam hi, thanks for your answers, i have met the same question about the NCCL, and his error happened when i used gpu devices named 1 and 2. But i try to train my network on other devices (e.g. 3 and 4) and it worked.  So i am so confused about that.

**Comment by amogkam** [comment]
@2a3b4c Is there a difference between the GPU devices? And is this using the latest version of Ray Lightning?

**Comment by 2a3b4c** [comment]
@amogkam  they are all GTX3090, i think there is no difference between, shall i need to run `nvidia-smi topo -m` to verify their topo? 
hmm, i used DDP to train my network, so i haven't use Ray Lightning? Is there a relationship between ray lightning and nccl?

**Comment by amogkam** [comment]
Oh this GitHub repo is the ray_lightning library. If you have issues with generic pytorch ddp, your best bet would be to raise an issue with PyTorch.

**Comment by 2a3b4c** [comment]
ok, thanks for your help!
